File: src/recursive/MaxSubarray.java
package recursive;

public class MaxSubarray {

	public static void main(String[] args) {
		int[] A = {13, -3, -25, 20, -3, -16, -23, 18, 20, -7, 12, -5, -22, 15, -4, 7};
		divideAndConquer(A);
	}
	
	public static void bruteForce(int[] A) {
		int sum = 0;
		int maxSum = Integer.MIN_VALUE;
		int start = 0;
		int end = 0;
		for (int i = 0; i < A.length; i++) {
			sum = 0;
			for (int j = i; j < A.length; j++) { // 注意单个元素也算子数组
				sum += A[j];
				if (sum > maxSum) { 
					maxSum = sum;
					start = i;
					end = j;
				}
				System.out.println(sum + " " + maxSum + " " + i + " " + j);
			}
		}
		System.out.println(start + " " + end);
	}

	public static void divideAndConquer(int[] A) {
		int[] max = divideAndConquer(A, 0, A.length - 1);
		System.out.println(max[0] + " " + max[1] + " " + max[2]);
	}

	private static int[] divideAndConquer(int[] A, int lo, int hi) { 
		if (hi == lo) { // hi 不可能小于 lo
			int[] ret = new int[3];
			ret[0] = A[lo];
			ret[1] = lo;
			ret[2] = hi;
			return ret;
		}
		int mid = (hi + lo) / 2; // floor

		int[] cross = maxCrossingSubarray(A, lo, mid, hi);
		int[] left = divideAndConquer(A, lo, mid);
		int[] right = divideAndConquer(A, mid + 1, hi);
		int[] ret = cross;
		if (left[0] > cross[0]) ret = left;
		if (right[0] > ret[0]) ret = right;
		return ret;
	}

	private static int[] maxCrossingSubarray(int[] A, int lo, int mid, int hi) {
		int max_left = Integer.MIN_VALUE;
		int max_right = Integer.MIN_VALUE;
		int start = 0;
		int end = 0;
		int sum = 0;
		for (int left = mid; left >= lo; left--) {
			sum += A[left];
			if (sum > max_left) {
                max_left = sum;
                start = left;
			}
		}
		sum = 0;
        for (int right = mid + 1; right <= hi; right++) {
        	sum += A[right];
        	if (sum >= max_right) {
        		max_right = sum;
        		end = right;
        	}
        }
        int max = max_left + max_right;
        int[] ret = new int[3];
        ret[0] = max;
        ret[1] = start;
        ret[2] = end;
        return ret; 
	}

}

File: src/leetcode/max_subarray.cc
#include <vector>
#include <limits>

class Solution {
    private:
        std::vector<int> A;

    public:
        Solution(std::vector<int> a) : A(a) {};

        // make sure that high > mid >= low
        std::vector<int> maxCrossingSubarray(int low, int high, int mid) {
            int left_sum = std::numeric_limits<int>::min();
            int sum = 0;
            int start = mid;
            int end = mid + 1;

            // i from mid to low 
            for (int i = mid; i >= low; i--) {
                sum += A[i];
                if (sum > left_sum) {
                    left_sum = sum;
                    start = i;
                }
            }

            int right_sum = std::numeric_limits<int>::min();
            sum = 0;
            // i from mid + 1 to high
            for (int i = mid + 1; i <= high; i++) {
                sum += A[i];
                if (sum > right_sum) {
                    right_sum = sum;
                    end = i;
                }
            }

            std::vector<int> ret;
            ret.push_back(right_sum + left_sum);
            ret.push_back(start);
            ret.push_back(end);
            return ret;
        }

        std::vector<int> maxSubarray(int low, int high) {
            if (low == high) {
                std::vector<int> max;
                max.push_back(A[low]);
                max.push_back(low);
                max.push_back(high);
                return max;
            }

            // high is bigger than low, mid is bigger than or equal with low
            int mid = (low + high) / 2;

            // max subarray containing mid and mid + 1
            // high >= mid + 1 > mid >= low
            std::vector<int> crossing = maxCrossingSubarray(low, high, mid);

            std::vector<int> left = maxSubarray(low, mid);
            std::vector<int> right = maxSubarray(mid + 1, high);

            // will influence the range
            // consider {1, 2, 3, -1, -2, -3, 6}
            std::vector<int> max = right;
            if (crossing[0] > max[0]) max = crossing;
            if (left[0] > max[0]) max = left;

            return max;
        }

};

int main(int argc, char const* argv[])
{   
    /* int arr[] = {1, 2, 3, 4, -5, -6, -7, -8}; */
    int arr[] = {1, 2, 3, -1, -2, -3, 6};
    std::vector<int> A(std::begin(arr), std::end(arr));

    int low = 0;
    int high = A.size() - 1;

    Solution solution(A);

    std::vector<int> result = solution.maxSubarray(low, high);

    for (std::vector<int>::iterator it = result.begin(); it != result.end(); it++) {
        printf("%d\n", *it);
    }
    return 0;
}
